### PR TITLE
[WIP] Added a lowering pass from FX -> NNC (also a decomposition pass)

### PR DESCRIPTION
**Pull request by Chillee**
Compiles the `DeepAndWide` model.

```
        num_features = 1000
        mod = DeepAndWide(num_features)

        # Phabricate sample inputs
        batch_size = 5
        embedding_size = 1000
        ad_emb_packed = torch.randn(batch_size, 1, embedding_size)
        user_emb = torch.randn(batch_size, 1, embedding_size)
        wide = torch.randn(batch_size, num_features)
        inps = (ad_emb_packed, user_emb, wide)
        out = torch.empty(5, 1)

        shape_env = ShapeProp(fx.symbolic_trace(mod)).propagate(*inps)
        mod = decompose(mod, shape_env)
        shape_env = ShapeProp(fx.symbolic_trace(mod)).propagate(*inps)
        cg = nnc_compile(mod, True, shape_env)

        iters = 100

        begin = time.time()
        for _ in range(iters):
            cg([ad_emb_packed, user_emb,wide, out])
        print("NNC time: ", time.time()-begin); begin = time.time()

        mod = mod
        for _ in range(iters):
            mod(*inps)
        print("PyTorch time: ", time.time()-begin); begin = time.time()
        print("Sums:", out.sum(), mod(*inps).sum())
----------------------
NNC time:  0.003288745880126953
PyTorch time:  0.046755075454711914
Sums: tensor(-152.5139) tensor(-152.5140)
```

### CODE REVIEW

File: torch/csrc/jit/tensorexpr/tensorexpr_init.cpp [hunk 1]
```diff
@@ -427,6 +430,14 @@ void initTensorExprBindings(PyObject* module) {
           "simplify",
           &tensorexpr::LoopNest::simplify,
           py::return_value_policy::reference)
+      .def(
+          "set_GPU_block_index",
```
Comment (bertmaher) [inline torch/csrc/jit/tensorexpr/tensorexpr_init.cpp:434]: micro-nit, I think usually our APIs use lowercase `gpu`

File: fx_compile.py [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+import torch.nn as nn
+import torch._C.te as te
+import torch.fx as fx
+from torch.fx.node import map_aggregate
+import time
+import numpy as np
+import pandas as pd
+import matplotlib.pyplot as plt
+import seaborn as sns
+import argparse
```
Comment (jamesr66a) [inline fx_compile.py:13]: nit: A bunch of unused inports

File: fx_compile.py [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+
+decomposition_rules[torch.addmm] = addmm_decompose
+
+def decompose(model: torch.nn.Module, shape_env) -> torch.nn.Module:
+    model = fx.symbolic_trace(model)
+    new_graph = fx.Graph()
+    env = {}
+    for node in model.graph.nodes:
+        if node.op == 'call_function' and node.target in decomposition_rules:
+            tracer = fx.proxy.GraphAppendingTracer(new_graph)
```
Comment (jamesr66a) [inline fx_compile.py:109]: Why is this needed? Doesn't `Proxy` use a `GraphAppendingTracer` by default

File: fx_compile.py [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+        ad_emb_packed = torch.randn(batch_size, 1, embedding_size)
+        user_emb = torch.randn(batch_size, 1, embedding_size)
+        wide = torch.randn(batch_size, num_features)
+        inps = (ad_emb_packed, user_emb, wide)
+        out = torch.empty(5, 1)
+
+        shape_env = ShapeProp(fx.symbolic_trace(mod)).propagate(*inps)
+        mod = decompose(mod, shape_env)
+        shape_env = ShapeProp(fx.symbolic_trace(mod)).propagate(*inps)
+        cg = nnc_compile(mod, True, shape_env)
```
Comment (jamesr66a) [inline fx_compile.py:419]: How bout `decompose` and `nnc_compile` do shape analysis internally to themselves? That way we're not symbolic tracing 4 times and adding more burden on the caller

File: fx_compile.py [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+
+############ New Stuff ##############
+
+def binary_mapping(op):
+    def f(a, b):
+        return op(a, b)
+    return f
+
+decomposition_rules = {}
+binary_decompositions = [
```
Comment (jamesr66a) [inline fx_compile.py:81]: I think for these rules to be applicable we would need to enforce the invariant that both of the arguments to these operators is Tensor. Maybe initially this can just be a contract with the user, but if this is going to be a "real" thing probably we'd want a more robust enforcement of that

File: fx_compile.py [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+
+def decompose(model: torch.nn.Module, shape_env) -> torch.nn.Module:
+    model = fx.symbolic_trace(model)
+    new_graph = fx.Graph()
+    env = {}
+    for node in model.graph.nodes:
+        if node.op == 'call_function' and node.target in decomposition_rules:
+            tracer = fx.proxy.GraphAppendingTracer(new_graph)
+            proxy_args = [fx.Proxy(env[x.name], tracer=tracer) if isinstance(x, fx.Node) else x for x in node.args]
+            new_node = decomposition_rules[node.target](*proxy_args).node
```
Comment (jamesr66a) [inline fx_compile.py:111]: For anything that's not an operator (including `addmm` here`, we might have a callsite with kwargs as well. I think this should section should look like:

```
proxy_args = map_arg(node.args, lambda n: Proxy(env[n.name]))
proxy_kwargs = map_arg(node.kwargs, lambda n: Proxy(env[n.name]))
new_node = decomposition_rules[node.target](*proxy_args, **proxy_kwargs).node
```

File: fx_compile.py [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+    inp_shapes = fx.node.map_aggregate(args, lambda arg: shape_env[arg.name] if isinstance(arg, fx.Node) else None)
+    return lowering_functions[op](name, shape_env[name][0], inp_shapes, nnc_args)
+
+def nnc_compile(model: torch.nn.Module, shape_specialize, shape_env) -> torch.nn.Module:
+    fx_model = fx.symbolic_trace(model)
+    env = {}
+    def get_te_shapes(name):
+        return [te.ExprHandle.int(i) for i in shape_env[name][0]]
+
+    def get_nnc_type(dtype):
```
Comment (jamesr66a) [inline fx_compile.py:329]: Seems like a function NNC should define/expose

File: fx_compile.py [hunk 8]
```diff
@@ -0,0 +1,432 @@
...
+    outs = None
+    inputs = []
+    module_attrs = []
+    for node in fx_model.graph.nodes:
+        if node.op == 'placeholder':
+            shapes = get_te_shapes(node.name)
+            env[node.name] = te.Placeholder(node.name, get_te_type(node.name), shapes)
+            inputs.append(env[node.name])
+        elif node.op == 'call_function':
+            result = lower_function(node.name, node.target, shape_env, lookup_env(node.args), node.args)
```
Comment (jamesr66a) [inline fx_compile.py:364]: Same sort of comment as above, this will miss callsites with some arguments as kwargs. If you're just handling functions, you might be able to use something like `inspect.Signature.bind()` to normalize args/kwargs

File: fx_compile.py [hunk 9]
```diff
@@ -0,0 +1,432 @@
...
+
+    outs = None
+    inputs = []
+    module_attrs = []
+    for node in fx_model.graph.nodes:
+        if node.op == 'placeholder':
+            shapes = get_te_shapes(node.name)
+            env[node.name] = te.Placeholder(node.name, get_te_type(node.name), shapes)
+            inputs.append(env[node.name])
+        elif node.op == 'call_function':
```
Comment (jamesr66a) [inline fx_compile.py:363]: I understand this is a prototype, but this also brings up the question of supporting methods (probably not too hard, but ideally we'd normalize those into functional forms) and modules (would we have to make all the modules functional and trace down into them? or have a canonical TE lowering for all of them, or something else)

**Review by jamesr66a — COMMENTED** [review]
Cool! Great to see results here, but it's pretty clear there's a long way to go to make this into a real thing

File: fx_compile.py (hunk 5 above)
Comment (Chillee) [reply fx_compile.py:81]: Doesn't this just need to be true of one of the arguments? If both of the arguments are not tensors then FX won't capture it.

Comment (jamesr66a) [reply fx_compile.py:81]: Ops/functions that return a non-Tensor output are one place where a non-Tensor proxy may appear. I think this shows up in pyper a few times

Comment (Chillee) [reply fx_compile.py:81]: Ah, well, if this is a problem we can just restrict it to operate only on tensors.

File: fx_compile.py (hunk 9 above)
Comment (Chillee) [reply fx_compile.py:363]: From "me thinking about it right now", I don't think modules need to be functional. Decomposing it would certainly make it easier to handle all of them, but you could also write TE lowerings without too much issue.

From my understanding, making modules functional is primarily only an issue for `autograd`, as that's the pass where the API needs access to module internals.

The main reason a "module functionalization" pass is needed is since you need the gradient of the output with respect to the module's weights. So if you had 

```
layer = nn.Linear(5,5)
def forward(x):
      return layer(x)
```

calling `grad(forward)(x)` gets you the gradient of `x` (the input), not `layer.weight`. What you would really want is `grad(forward)(x, layer.weight)`, but that's not really possible right now for a variety of reasons.

For a lowering pass, or for vmap, there's no issues with just accessing the module attributes and decomposing it into

```
def forward(x):
    return x @ layer.weight + layer.bias
```

You can see that in the current DeepAndWide there's module attributes I'm accessing that are getting "compiled".

Comment (jamesr66a) [reply fx_compile.py:363]: My comment was more about maintainability in the codebase. The way I see it there are two options:

* Have a separate canonical TE lowering for each standard library `Module` instance. This is prohibitive and probably not a good idea, as each `Module` will need to be written twice and those implementations will need to stay in sync.
* Have the module written once and automatically lowerable to TE. I think there are sub-possibilities here:
  a. Make it so that the `Modules` are convertible to TE, such as via symbolic tracing + a lowering on the RISCy primitives. This would potentially require modification of the way we currently write `Modules` so that they would be captureable. Specifically, for them to be captureable in FX they would need to be functional (in the sense of the ops are functional, not the sense that they don't have attributes).
  b. Make it so that the `Module`s are written in TE in the first place. This would be a huge change and potentially not in the spirit of being Pythonic, but maybe things will go fast.

Comment (Chillee) [reply fx_compile.py:363]: Ah, I see what you mean. Yes, ideally we would make it so that `Modules` are convertible to TE. I think your work showed that this isn't actually that bad, especially if we're shape specializing.

There might be a couple we need special handling for, but overall it seems acceptable.

**Comment by jamesr66a** [comment]
Please move this to a PR against the examples repo: https://github.com/pytorch/examples/tree/master/fx

Thanks!
